File: Data-Structure-I/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(),nums.end());
        int n = nums.size();
        int a;
        for(int i = 0; i<n;i++){
            if(i == 0 || (i>0 && nums[i] != nums[i-1])){
                a = nums[i];
                int low = i+1 , high = n-1;
                while(low<high){
                    if(nums[low] + nums[high] > -a) high--;
                    else if(nums[low] + nums[high] < -a) low++;
                    else{
                        ans.push_back({a,nums[low],nums[high]});
                        while(low<high && nums[low] == nums[low+1]) low++;
                        while(low<high && nums[high] == nums[high-1]) high--;
                        low++; high--;
                    }
                }
            }
        }
        return ans;
    }
};

File: Data-Structure-I/15-3sum.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function threeSum($nums) {
        sort($nums);
        $count = count($nums);
        $ans = [];
        for ($index = 0 ; $index <$count ; $index ++) {
            $number = $nums[$index];
            if( $index>0 && $number == $nums[$index-1]) {
                continue;
            }
            $low = $index+1;
            $high = count($nums)-1;
            while ($low<$high)
            {
                $threeSum = $number + $nums[$low] + $nums[$high];
                if($threeSum>0)
                {
                    --$high;
                }

                elseif($threeSum<0) {
                    ++$low;
                }
                else{
                    $ans [] = [$number,$nums[$low],$nums[$high]];
                    ++$low;
                    while (isset($nums[$low],$nums[$low-1])&&$nums[$low] == $nums[$low-1])
                    {
                        ++$low;
                    
                    }
                }
            }
        }
        return $ans;
    }
}
